### PR TITLE
trigger backfill in reversed order

### PR DESCRIPTION
**Pull request by honnix**
minimum effort to make this feature. everything being the same, but to
trigger backfill in reversed order, user only needs to put `start` after
`end`, where `start` is inclusive and `end` is exclusive as it is today.

### CODE REVIEW

**Comment by honnix** [comment]
@ulzha wdyt of this minimum effort? The user facing side is, if `start` is after `end`, it is reversed.

File: styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java [hunk 1]
```diff
@@ -122,17 +125,53 @@ public static int instancesInRange(Instant firstInstant, Instant lastInstant, Sc
...
     }
-    return number;
+
+    return listOfInstants;
+  }
+
+  /**
+   * Gets the number of instants between firstInstant (inclusive) and lastInstant (exclusive)
+   * according to the {@link Schedule}. This works in a reversed order, meaning firstInstant should
+   * be after lastExecution.
```
Comment (ulzha) [inline styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java]: Typo, `lastInstant`?

**Comment by ulzha** [comment]
A bit worried that the appearance of these backfills in UI will be confusing, people will be expecting the first chronological instance in START column in `styx backfill ls`...

File: styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java (hunk 1 above)
Comment (honnix) [reply styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java]: Seems like yes.

**Comment by honnix** [comment]
That's hard to say. I would expect the reversed order as I exactly asked. :)

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java [hunk 2]
```diff
@@ -281,8 +282,10 @@ private static boolean isActiveState(RunStateData runStateData) {
           Status.BAD_REQUEST.withReasonPhrase("end parameter not aligned with schedule"));
     }
 
+    final boolean reversed = input.start().isAfter(input.end());
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: Can this check move into `instants()` ?

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java [hunk 3]
```diff
@@ -347,8 +350,10 @@ private static boolean isActiveState(RunStateData runStateData) {
       throw new RuntimeException(e);
     }
 
-    final List<Instant> processedInstants = rangeOfInstants(
-        backfill.start(), backfill.nextTrigger(), backfill.schedule());
+    final boolean reversed = backfill.start().isAfter(backfill.end());
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: same as above

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 4]
```diff
@@ -342,6 +352,32 @@ public void shouldGetBackfillStatusWithStatusByDefault() throws Exception {
     assertJson(response, "statuses.active_states", hasSize(24));
   }
 
+  @Test
+  public void shouldGetBackfillStatusWithStatusByDefaultReversed() throws Exception {
+    sinceVersion(Api.Version.V3);
+
+    WorkflowInstance wfi = WorkflowInstance.create(BACKFILL_2.workflowId(), "2017-01-01T23");
```
Comment (danielnorberg) [inline styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: storage should be mocked out in these tests

Comment (danielnorberg) [reply styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: Maybe there's a bit too much event replay logic in `BackfillResource` itself. Factoring out that should ease testing. I.e., it should be possible to mock `getBackfillRunState()` in `BackfillResourceTest` and test `getBackfillRunState()` in a separate `ReplayEventsTest`.

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java (hunk 3 above)
Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: It wouldn't change the logic, but would change the semantics a bit though. Instead of checking only `start` and `end`, that would check `start`, `nextTrigger`, and `end`.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java (hunk 4 above)
Comment (danielnorberg) [reply styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: can be another PR I guess

File: styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java [hunk 5]
```diff
@@ -104,15 +106,16 @@ public static Instant nextInstant(Instant instant, Schedule schedule) {
   }
 
   /**
-   * Gets the number of workflow instances between firstInstant (inclusive) and lastInstant (exclusive)
-   * according to the workflow's schedule.
+   * Gets the number of instants between firstInstant (inclusive) and lastInstant (exclusive)
```
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java]: Nitpick: It actually gets an ordered list of the instants, not the number.

File: styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java [hunk 6]
```diff
@@ -122,17 +125,53 @@ public static int instancesInRange(Instant firstInstant, Instant lastInstant, Sc
     }
 
     final ExecutionTime executionTime = ExecutionTime.forCron(cron(schedule));
+    final List<Instant> listOfInstants = new ArrayList<>();
```
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java]: Nitpick: hungarian notation? I would just call this `instants` =)

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java [hunk 7]
```diff
@@ -21,12 +21,16 @@
 package com.spotify.styx.util;
 
 import static com.spotify.styx.util.TimeUtil.addOffset;
-import static com.spotify.styx.util.TimeUtil.instancesInRange;
+import static com.spotify.styx.util.TimeUtil.instantsInRange;
+import static com.spotify.styx.util.TimeUtil.instantsInReversedRange;
 import static com.spotify.styx.util.TimeUtil.isAligned;
 import static com.spotify.styx.util.TimeUtil.lastInstant;
 import static com.spotify.styx.util.TimeUtil.nextInstant;
+import static java.time.Instant.parse;
```
Comment (danielnorberg) [inline styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: This static import produces a lot of noise in this file, makes it harder to check what actually changed.

Comment (danielnorberg) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Although the github diff highlights the actual changes, so no big deal

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java [hunk 8]
```diff
@@ -187,53 +191,149 @@ public void shouldAddOffsetWithWeek() {
...
-    final Instant firstTimeWeeks = Instant.parse("2016-01-11T00:00:00.00Z");
-    final Instant lastTimeWeeks = Instant.parse("2016-01-18T00:00:00.00Z");
-    final Instant firstTimeMonths = Instant.parse("2010-01-01T00:00:00.00Z");
-    final Instant lastTimeMonths = Instant.parse("2016-01-01T00:00:00.00Z");
-
-    assertThat(instancesInRange(firstTimeHours, lastTimeHours, Schedule.HOURS), is(9));
-    assertThat(instancesInRange(firstTimeDays, lastTimeDays, Schedule.DAYS), is(9));
-    assertThat(instancesInRange(firstTimeWeeks, lastTimeWeeks, Schedule.WEEKS), is(1));
-    assertThat(instancesInRange(firstTimeMonths, lastTimeMonths, Schedule.YEARS), is(6));
+    final Instant firstTimeHours = parse("2016-01-19T00:00:00.00Z");
```
Comment (danielnorberg) [inline styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Nitpick: `NumberOf` in the names of all these tests should now be removed? E.g. `shouldGetCorrectNumberOfInstants` becomes `shouldGetCorrectInstants`.

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java [hunk 9]
```diff
@@ -187,53 +191,149 @@ public void shouldAddOffsetWithWeek() {
...
+  public void shouldReturnEmptyList() {
+    final Instant firstTimeHours = parse("2016-01-19T00:00:00.00Z");
+    final Instant lastTimeHours = parse("2016-01-19T00:00:00.00Z");
+
+    assertTrue(
+        instantsInRange(firstTimeHours, lastTimeHours, Schedule.parse("0 * * * *")).isEmpty());
   }
 
   @Test(expected = IllegalArgumentException.class)
   public void shouldGetExceptionIfLastInstantIsBeforeFirstInstant() {
```
Comment (danielnorberg) [inline styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Maybe not part of this PR but this use of `@Test(expected = IllegalArgumentException.class)` might produce false positives for these tests if `parse` throws.

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java [hunk 10]
```diff
@@ -187,53 +191,149 @@ public void shouldAddOffsetWithWeek() {
...
+    final Instant firstTimeHours = parse("2016-01-19T03:00:00.00Z");
+    final Instant lastTimeHours = parse("2016-01-19T00:00:00.00Z");
+    final Instant firstTimeDays = parse("2016-01-13T00:00:00.00Z");
+    final Instant lastTimeDays = parse("2016-01-10T00:00:00.00Z");
+    final Instant firstTimeWeeks = parse("2016-01-18T00:00:00.00Z");
+    final Instant lastTimeWeeks = parse("2016-01-11T00:00:00.00Z");
+    final Instant firstTimeMonths = parse("2013-01-01T00:00:00.00Z");
+    final Instant lastTimeMonths = parse("2010-01-01T00:00:00.00Z");
+
+    assertThat(instantsInReversedRange(firstTimeHours, lastTimeHours, Schedule.HOURS),
```
Comment (danielnorberg) [inline styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Is it expected that `instantsInRange(x, y, ...)` and `instantsInReversedRange(y, x, ...)` do not produce the same set of instants?

If the user supplies the same start/end timestamps to the backfill create command but in reverse, should it not produce the exact same workflow instances but just in reverse?

File: styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java [hunk 11]
```diff
@@ -104,15 +106,16 @@ public static Instant nextInstant(Instant instant, Schedule schedule) {
...
+   * according to the {@link Schedule}.
    *
    * @param firstInstant The first instant
    * @param lastInstant  The last instant
    * @param schedule     The schedule of the workflow
-   * @return the number of instances within the range
+   * @return instants within the range
    */
-  public static int instancesInRange(Instant firstInstant, Instant lastInstant, Schedule schedule) {
+  public static List<Instant> instantsInRange(Instant firstInstant, Instant lastInstant,
```
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java]: Check that `firstInstant.isBefore(lastInstant)` ?

File: styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java [hunk 12]
```diff
@@ -122,17 +125,53 @@ public static int instancesInRange(Instant firstInstant, Instant lastInstant, Sc
...
+   * Gets the number of instants between firstInstant (inclusive) and lastInstant (exclusive)
+   * according to the {@link Schedule}. This works in a reversed order, meaning firstInstant should
+   * be after lastInstant.
+   *
+   * @param firstInstant The first instant
+   * @param lastInstant  The last instant
+   * @param schedule     The schedule of the workflow
+   * @return instants within the range
+   */
+  public static List<Instant> instantsInReversedRange(Instant firstInstant, Instant lastInstant,
```
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/util/TimeUtil.java]: check that `firstInstant.isAfter(lastInstant)` ?

File: styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java [hunk 13]
```diff
@@ -379,4 +380,9 @@ private static boolean isActiveState(RunStateData runStateData) {
 
     return Stream.concat(processedStates.stream(), waitingStates.stream()).collect(toList());
   }
+
+  private static List<Instant> instants(Instant start, Instant end, Schedule schedule) {
+    return start.isAfter(end) ? instantsInReversedRange(start, end, schedule) :
```
Comment (danielnorberg) [inline styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: In order to always return the same set of instants, maybe this should be:

```
return start.isAfter(end))
  ? Lists.reverse(instantsInRange(end, start, schedule))
  : instantsInRange(start, end, schedule);
```

If not, at least please format the ternary conditional operator like this (when it does not fit on one line):
```
return start.isAfter(end) 
  ? instantsInReversedRange(start, end, schedule) 
  : instantsInRange(start, end, schedule);
```

Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java]: It will be different because of inclusive and exclusive. Will fix the format.

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java (hunk 10 above)
Comment (honnix) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: It is expected due to inclusive (x) and exclusive (y).

Comment (danielnorberg) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Yeah, and I think that might be quite confusing for the user. Might it be easier to use if `styx backfill create <component> <workflow> 2017-01-01 2018-01-01 <concurrency>` and 
`styx backfill create <component> <workflow> 2018-01-01 2017-01-01 <concurrency>` result in the same partitions, just the order reversed?

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java [hunk 14]
```diff
@@ -244,29 +245,44 @@ public void shouldReturnEmptyList() {
...
 
-  @Test(expected = IllegalArgumentException.class)
+  @Test
   public void shouldGetExceptionIfLastInstantIsBeforeFirstInstant() {
     final Instant firstTimeHours = parse("2016-01-19T10:00:00.00Z");
     final Instant lastTimeHours = parse("2016-01-19T09:00:00.00Z");
-    instantsInRange(firstTimeHours, lastTimeHours, Schedule.HOURS);
+    try {
+      instantsInRange(firstTimeHours, lastTimeHours, Schedule.HOURS);
+      fail();
```
Comment (danielnorberg) [inline styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Use `ExpectedException` rule

Comment (honnix) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Thought about this. What is the guideline for using this rule? In case it is quite a bit code achieving the same goal.

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java (hunk 10 above)
Comment (honnix) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: I think it might be OK because we have been explicitly stating the `end` is exclusive, so unless we want to make both of them inclusive for normal and reversed order, it might be more confusing.

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java (hunk 14 above)
Comment (honnix) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: nvm, it is worth it.

File: styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java (hunk 10 above)
Comment (danielnorberg) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: Maybe this needs some user testing.

Comment (danielnorberg) [reply styx-common/src/test/java/com/spotify/styx/util/TimeUtilTest.java]: The more I think about this, the more it seems like having an explicit `reverse` flag might be better, both for humans using the cli and for machines calling the api.

E.g. if we add backfill creation to System-Z and then have a `reverse` check box that works by swapping the dates and shifting them 1 partition in one direction, that seems awkward to me. Then I would rather just have a `reverse=true|false` parameter.

And `styx backfill create --reverse ...` seems more clear than swapping start/end.

File: styx-client/src/test/java/com/spotify/styx/client/StyxApolloClientTest.java [hunk 15]
```diff
@@ -206,6 +206,43 @@ public void shouldCreateBackfill() throws Exception {
...
+    final BackfillInput backfillInput = BackfillInput.newBuilder()
+        .start(start)
+        .end(end)
+        .component("foo-comp")
+        .workflow("bar-wf")
+        .concurrency(1)
+        .reverse(true)
+        .build();
+
+    final Backfill backfill = Backfill.newBuilder()
```
Comment (NatashaL) [inline styx-client/src/test/java/com/spotify/styx/client/StyxApolloClientTest.java]: The values for `backfillInput` and this `backfill` seem magic. Either extract them, or use them once and then reuse them, like done here https://github.com/spotify/styx/pull/455/files?utf8=%E2%9C%93&diff=split#diff-ae03c6a48f8fa80596ad167c226fdde5R272

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java [hunk 16]
```diff
@@ -158,7 +160,8 @@ boolean triggerNextPartitionAndProgress(StorageTransaction tx,
                                           String id,
                                           Workflow workflow,
                                           Instant initialNextTrigger,
-                                          int remainingCapacity) {
+                                          int remainingCapacity,
+                                          boolean reversed) {
     final Backfill momentBackfill = tx.backfill(id).orElseThrow(() ->
```
Comment (NatashaL) [inline styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: I understand this is older, but why `momentBackfill`? Why not just backfill?

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java [hunk 17]
```diff
@@ -169,13 +172,13 @@ boolean triggerNextPartitionAndProgress(StorageTransaction tx,
 
     final Instant momentNextTrigger = momentBackfill.nextTrigger();
```
Comment (NatashaL) [inline styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: Same as before, why use `moment` ?

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java [hunk 18]
```diff
@@ -223,11 +226,45 @@ boolean triggerNextPartitionAndProgress(StorageTransaction tx,
...
+
+  private static Instant getNextPartition(Backfill momentBackfill,
+                                   Instant momentNextTrigger,
+                                   boolean reversed) {
+    return reversed
+        ? previousInstant(momentNextTrigger, momentBackfill.schedule())
+        : nextInstant(momentNextTrigger, momentBackfill.schedule());
+  }
+
+  private static boolean capacityReached(Backfill momentBackfill, Instant initialNextTrigger,
```
Comment (NatashaL) [inline styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: As far as code-style goes, I don't mind using either ternary of if-else, but let's stick to one and the same, where it applies. `exclusiveEndTrigger`, `getNextPartition` and `capacityReached` can all be written using the same notation

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java (hunk 16 above)
Comment (honnix) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: Just to make it clear this is a fresh backfill read from database at this moment.

File: styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java [hunk 19]
```diff
@@ -395,6 +432,36 @@ public void shouldPostBackfill() throws Exception {
...
+    assertThat(postedBackfill.id().matches("backfill-[\\d-]+"), is(true));
+    assertThat(postedBackfill.start(), equalTo(Instant.parse("2017-01-01T00:00:00Z")));
+    assertThat(postedBackfill.end(), equalTo(Instant.parse("2017-02-01T00:00:00Z")));
+    assertThat(postedBackfill.workflowId(), equalTo(WorkflowId.create("component", "workflow2")));
+    assertThat(postedBackfill.concurrency(), equalTo(1));
+    assertThat(postedBackfill.description(), equalTo(Optional.empty()));
+    assertThat(postedBackfill.nextTrigger(), equalTo(Instant.parse("2017-01-31T23:00:00Z")));
+    assertThat(postedBackfill.schedule(), equalTo(Schedule.HOURS));
+    assertThat(postedBackfill.allTriggered(), equalTo(false));
+    assertThat(postedBackfill.halted(), equalTo(false));
```
[the remark this answers is not in the document]
Comment (honnix) [reply styx-api-service/src/test/java/com/spotify/styx/api/BackfillResourceTest.java]: this has been fixed.

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java (hunk 16 above)
Comment (honnix) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: It was named like this when both the `old` backfill and the one just read were in the same method, and stayed unchanged over refactorings. It is totally ok to rename if it creates confusion now.
